File: leetcode/735.行星碰撞.py
#
# @lc app=leetcode.cn id=735 lang=python3
#
# [735] 行星碰撞
#

from typing import List


# @lc code=start
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        res = []
        i = 0
        n = len(asteroids)
        while i < n:
            asteroid = asteroids[i]
            if not res or not (
                res[-1] > 0 and asteroid < 0
            ):
                res.append(asteroid)
                i += 1
                continue
            if abs(asteroid) > res[-1] > 0:
                res.pop()
                continue
            if abs(asteroid) == abs(res[-1]):
                res.pop()
            i += 1

        return res


# @lc code=end
print(Solution().asteroidCollision([5, 10, -5]))
print(Solution().asteroidCollision([8, -8]))
print(Solution().asteroidCollision([10, 2, -5]))
print(Solution().asteroidCollision([-2, -1, 1, 2]))
print(Solution().asteroidCollision([-2, -2, 1, -2]))

File: leetcode/735.行星碰撞.cpp
#include <iostream>
#include <vector>
using namespace std;
/*
 * @lc app=leetcode.cn id=735 lang=cpp
 *
 * [735] 行星碰撞
 */

// @lc code=start
class Solution
{
public:
    vector<int> asteroidCollision(vector<int> &asteroids)
    {
        vector<int> res;
        int i = 0;
        int n = asteroids.size();
        while (i < n)
        {
            int asteroid = asteroids[i];
            if (res.empty() || !(res.back() > 0 && asteroid < 0))
            {
                res.push_back(asteroid);
                i++;
                continue;
            }
            if ((-asteroid) > res.back())
            {
                res.pop_back();
                continue;
            }
            if ((-asteroid) == res.back())
            {
                res.pop_back();
            }
            i++;
        }
        return res;
    }
};
// @lc code=end
int main()
{
    Solution s;
    int arr[] = {5, 10, -5};
    int n = sizeof(arr) / sizeof(arr[0]);
    vector<int> asteroids(arr, arr + n);
    vector<int> res = s.asteroidCollision(asteroids);
    for (auto &&i : res)
    {
        cout << i << ",";
    }
    cout << endl;
}
